### PR TITLE
Explore: Improve parsing ranges from URL 

### PR DESCRIPTION
**Pull request by ifrost**
Related to: #72497

Please check #72497 and #72578 for more details.

The main change is switching the type from:

```typescript
/** @internal */
export interface ExploreUrlState<T extends DataQuery = AnyQuery> {
  datasource: string | null;
  queries: T[];
  range: RawTimeRange;
  panelsState?: ExplorePanelsState;
}
```

to

```typescript
// enforce type-incompatibility with RawTimeRange to ensure it's parsed and converted.
// URLRangeValue may be a string representing UTC time in ms, which is not a compatible
// value for RawTimeRange when used as a string (it could only be an ISO formatted date)
export type URLRangeValue = string | { __brand: 'URL Range Value' };

/**
 * @internal
 */
export type URLRange = {
  from: URLRangeValue;
  to: URLRangeValue;
};

/** @internal */
export interface ExploreUrlState<T extends DataQuery = AnyQuery> {
  datasource: string | null;
  queries: T[];
  range: URLRange;
  panelsState?: ExplorePanelsState;
}
```

This is to enforce the need to convert the range provided in the URL into RawTimeRange.

The rest of the changes is fixing created this way type errors and amending tests to ensure that conversion happens.

This will also:
* improve handling malformed from/to values and fallback to default range when the conversion is not successful (e.g. passing incorrect string as a param, e.g. `from=boom`).
* unify how data links in Explore generate URLs with absolute ranges (previously it would stringify RawTimeRange ending up with ISO dates in from/to, Explore can handle it but it makes it consistent with how absolute URLs are generated by Explore itself) 

Note: 

* Explore/Dashboards save from/to values in the URL as a string ([Dashboards code](https://github.com/grafana/grafana/blob/main/public/app/features/dashboard/services/TimeSrv.ts#L330), [Explore code](https://github.com/grafana/grafana/blob/v10.1.x/public/app/features/explore/hooks/useStateSync/index.ts#L397)). Absolute dates are stored to a string containing milliseconds.
* When milliseconds as a strings are passed in the URL and the code assumes that they are the same as RawTimeRange strage things start to happen. Such string is passed in Explore [here](https://github.com/grafana/grafana/blob/v10.1.x/public/app/features/explore/state/time.ts#L80). It may end up as unparsed string in moment.js [here](https://github.com/grafana/grafana/blob/v10.1.x/packages/grafana-data/src/datetime/parser.ts#L75) and marked as invalid as moment.js requires ISO formatted dates passed as strings.
* The above missing conversion happens only when time range is updated in the URL (when history changes) because initialize Explore handles this indirectly [here](https://github.com/grafana/grafana/blob/v10.1.x/public/app/features/explore/state/explorePane.ts#L147C16-L147C24) (getRange does additional parsing, but we need to for updateTime from URL as well). 

### Deprecation notice

Passing `range` to `mapInternalLinkToExplore` is now optional. Previously to skip passing the range consumers had to pass it as an empty object (e.g. `{} as any}). Passing an empty object is now deprecated and `undefined` should be used instead. After deprecation period, if not amended, it will lead to runtime errors. The deprecation period is set to 1 year.

### CODE REVIEW

File: packages/grafana-data/src/field/fieldOverrides.test.ts [hunk 1]
```diff
@@ -802,8 +802,8 @@ describe('getLinksSupplier', () => {
 
     const links = supplier({ valueRowIndex: 0 });
     const rangeStr = JSON.stringify({
-      from: range.from.toISOString(),
-      to: range.to.toISOString(),
```
Comment (Elfo404) [inline packages/grafana-data/src/field/fieldOverrides.test.ts:806]: why is this needed?

File: packages/grafana-data/src/types/explore.ts [hunk 2]
```diff
@@ -1,15 +1,28 @@
...
 import { PreferredVisualisationType } from './data';
-import { RawTimeRange, TimeRange } from './time';
+import { TimeRange } from './time';
 
 type AnyQuery = DataQuery & Record<string, any>;
 
+// enforce type-incompatibility with RawTimeRange to ensure it's parsed and converted.
+// URLRangeValue may be a string representing UTC time in ms, which is not a compatible
+// value for RawTimeRange when used as a string (it could only be an ISO formatted date)
+export type URLRangeValue = string | { __brand: 'URL Range Value' };
```
Comment (Elfo404) [inline packages/grafana-data/src/types/explore.ts:11]: what is the purpose of `{__brand: 'URL Range Value' }`?

File: public/app/features/explore/hooks/useStateSync/index.ts [hunk 3]
```diff
@@ -358,7 +358,7 @@ const urlDiff = (
 } => {
   const datasource = !isEqual(currentUrlState?.datasource, oldUrlState?.datasource);
   const queries = !isEqual(currentUrlState?.queries, oldUrlState?.queries);
-  const range = !areRangesEqual(currentUrlState?.range || DEFAULT_RANGE, oldUrlState?.range || DEFAULT_RANGE);
+  const range = !isEqual(currentUrlState?.range || DEFAULT_RANGE, oldUrlState?.range || DEFAULT_RANGE);
```
Comment (Elfo404) [inline public/app/features/explore/hooks/useStateSync/index.ts:361]: this changes the behavior when the properties in the range objects are not in the same order, are we ok with it?

File: packages/grafana-data/src/types/explore.ts (hunk 2 above)
Comment (ifrost) [reply packages/grafana-data/src/types/explore.ts:11]: It makes it impossible to cast URLRange directly to RawTimeRange:

before (no type error when converting from URLRange to RawTimeRange):

Screenshot 2023-08-16 at 12 32 13

after:

Screenshot 2023-08-16 at 12 32 28

File: public/app/features/explore/hooks/useStateSync/index.ts (hunk 3 above)
Comment (ifrost) [reply public/app/features/explore/hooks/useStateSync/index.ts:361]: I think lodash ignores the order of properties:

```typescript
import _ from 'lodash'
    
let a = {};
a.from = 1;
a.to = 2;

let b = {}
b.to = 2;
b.from = 1;

console.log("a =", a);
console.log("b =", b);
console.log("_.isEqual(a,b)", _.isEqual(a,b));
```

Console:

```
a =(2) {from: 1, to: 2}
b =(2) {to: 2, from: 1}
_.isEqual(a,b) true
```

File: packages/grafana-data/src/field/fieldOverrides.test.ts (hunk 1 above)
Comment (ifrost) [reply packages/grafana-data/src/field/fieldOverrides.test.ts:806]: This makes internal URLs generated by data links consistent with Explore URLs (uses ms instead of ISO)

**Review by Elfo404 — APPROVED** [review]
